### ISSUE TITLE
Does we have any docs of how to create new grammar/rules for inverse text normalization for a new language?

### ISSUE THREAD

**Issue by lociko** [NVIDIA/NeMo#2742]
**Is your feature request related to a problem? Please describe.**

I want to create ITN grammar for the Ukrainian language. As we already have one for Russian I think it would be easy to create a new one for Ukrainian (as it has almost the same rules a Russian).

**Describe the solution you'd like**

I'm looking for some docs or guides on how to create a new grammar for Inver Text Normalization for a new language?

**Describe alternatives you've considered**

**Additional context**

**Comment by ekmb** [comment]
The ITN rules for Ru are based on non-deterministic TN Rules. ITN grammars are just an inverse of the TN rules.
The easiest way to start would be to copy both TN and ITN Ru folders (note, import paths should be updated, e.g., `from nemo_text_processing.text_normalization.ru.utils import get_abs_path` -> from nemo_text_processing.text_normalization.ukr.utils import get_abs_path). 
Then, you would need to modify data files (https://github.com/NVIDIA/NeMo/tree/main/nemo_text_processing/text_normalization/ru/data) and alphabet (https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/text_normalization/ru/alphabet.py) files to use Ukrainian instead of Russian.

**Comment by lociko** [comment]
Thanks! It looks like work, but I can see some problems when numbers should be in of the genders, cases, or plural forms. how NeMo chose the case, gender, or plural for the Russian language?

**Comment by lociko** [comment]
And maybe one more question: why does one number in a case has more than one different form? 
For example in https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/text_normalization/ru/data/numbers/2_cardinals_genitive_%D1%80%D0%BE%D0%B4%D0%B8%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%B9.tsv

the 700 has 5 different variations (some of them is not belongs to genitive case as I see):
```
700	семисот
700	семистам
700	семистах
700	семьсот
700	семьюстами
```
in other cases it has only one representation, so what is the rule to chose different forms? I just want to understand the approach of how was the Russian grammar designed to transfer this knowledge to Ukrainian.

Thanks!

**Comment by ekmb** [comment]
> Thanks! It looks like work, but I can see some problems when numbers should be in of the genders, cases, or plural forms. how NeMo chose the case, gender, or plural for the Russian language?

Yes, for TN we provide all possible options, so gender and case sometimes don't match the context. The TN is non-deterministic. However, this is not an issue for the ITN task. All multiple TN options will be mapped to a single (shortest path) option for the ITN task.

**Comment by ekmb** [comment]
> And maybe one more question: why does one number in a case has more than one different form?
> For example in https://github.com/NVIDIA/NeMo/blob/main/nemo_text_processing/text_normalization/ru/data/numbers/2_cardinals_genitive_%D1%80%D0%BE%D0%B4%D0%B8%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%B9.tsv
> 
> the 700 has 5 different variations (some of them is not belongs to genitive case as I see):
> 
> ```
> 700	семисот
> 700	семистам
> 700	семистах
> 700	семьсот
> 700	семьюстами
> ```
> 
> in other cases it has only one representation, so what is the rule to chose different forms? I just want to understand the approach of how was the Russian grammar designed to transfer this knowledge to Ukrainian.
> 
> Thanks!

that's a bug, only a single form of the word should be present, I'll send a PR with the fix. Thanks!

**Comment by liaeh** [comment]
Piggy-backing on this, I would also be interested in creating an ITN grammar, but for French. Since there is no TN grammar for French in NeMo to inverse, what would be the best approach here? 

Thanks for the tips 😄

**Comment by liaeh** [comment]
Another question - how were the weights determined in the en-itn? Looking through the implementation, I find some lines where weights are hard-coded, e.g.:
https://github.com/NVIDIA/NeMo/blob/0c7fbad2903f0a4bd5eba0e959d9563b365e3346/nemo_text_processing/inverse_text_normalization/en/taggers/money.py#L86

How is the value of these weights determined?

**Comment by xingchensong** [comment]
also interested in weights designing.

**Comment by ekmb** [comment]
The weights are determined experimentally. The weights are needed when there is a union operation. To set a preference for a particular option, we use a small negative weight. 

@liaeh there is an open [PR for French ITN grammars](https://github.com/NVIDIA/NeMo/pull/2921), the PR is under review now.

The easiest way to start adding grammar for a new language would be to come up with a list of test cases for the target language and then use the rules for the existing languages as a reference. The [`data` folder](https://github.com/NVIDIA/NeMo/tree/main/nemo_text_processing/inverse_text_normalization/en/data) entries should be adjusted for the target language. A detailed tutorial on how to add grammar for a new language is on our roadmap.

**Comment by xingchensong** [comment]
> The weights are determined experimentally. The weights are needed when there is a union operation. To set a preference for a particular option, we use a small negative weight.
> 
> @liaeh there is an open [PR for French ITN grammars](https://github.com/NVIDIA/NeMo/pull/2921), the PR is under review now.
> 
> The easiest way to start adding grammar for a new language would be to come up with a list of test cases for the target language and then use the rules for the existing languages as a reference. The [`data` folder](https://github.com/NVIDIA/NeMo/tree/main/nemo_text_processing/inverse_text_normalization/en/data) entries should be adjusted for the target language. A detailed tutorial on how to add grammar for a new language is on our roadmap.

Hi, are there any tricks for weight-tuning to share? I'm working on developing Chinese ITN grammars and graph weight has a great influence on the performance of classification : (

**Comment by xingchensong** [comment]
Thanks !
